OS/Atmega32/src/SOS.c

```diff
@@ -0,0 +1,441 @@
+#include "SOS.h"
+
+/*- LOCAL MACROS
+------------------------------------------*/
+#define  SECOND_TO_MICROSECOND      (uint32_t)( 1000000 )
+#define  SYSTICK_OVERFLOW_FACTOR    (uint16_t)( 256 )
+#define  IDLE_TASK                  (uint8_t)(255)
+#define  SLEEP_ENABLE               (uint8_t)(0x80)
+#define  SLEEP_BITS_MASK            (uint8_t)(0x70)
+#define  IDLE_MODE                  (uint8_t)(0x00)
+
+#define  CALCULATE_CPU_LOAD         (1)
+
+#if CALCULATE_CPU_LOAD
+   #define  CPU_IDLE                (uint8_t)(0)
+   #define  CPU_WORKING             (uint8_t)(1)  
+   #define  CPU_CALCULATION_WINDOW  (uint64_t)(200)      
+#endif
+
+/*- LOCAL FUNCTIONS PROTOTYPES
+----------------------------*/
+#if CALCULATE_CPU_LOAD
+   void TransmitCPULoad(void);
+#endif
+void Systick_ISR(void);
+static void Idle_Task(void);
+static void SOS_UpdateStates(void);
+static ERROR_STATE_t GetFreqPrescaleRatio(float32_t * Ratio);
+
+/*- GLOBAL STATIC VARIABLES
+-------------------------------*/
+static STR_Task_t gastr_Tasks[MAX_TASKS_NUMBER];
+static uint8_t gu8_CreatedTasks = 0;
+static uint64_t gu64_Ticks = 1;
+static uint64_t gu64_SYSTICK_OverflowsNeeded = 0;
+static uint64_t gu64_SYSTICK_Overflows = 0;
+
+#if CALCULATE_CPU_LOAD
+   static uint8_t CPU_State = CPU_IDLE;
+   static uint64_t CPU_IdleTime = 0;
+   static uint64_t CPU_WorkingTime = 0;
+   static uint8_t CPU_Load = 0;
+#endif
+
+/*- GLOBAL EXTERN VARIABLES
+-------------------------------*/
+extern const STR_SYSTICK_config_t gstr_SYSTICK_Config;
+extern const uint16_t gu16_TICK_Interval;
+
+/*- LOCAL FUNCTIONS IMPLEMENTATION
+------------------------*/
+#if CALCULATE_CPU_LOAD
+   #include "UART.h"
+   void TransmitCPULoad(void)
+   {
+      UART_Init(UART_CH_0);
+      CPU_Load = (uint8_t)((CPU_WorkingTime*1.0)/(CPU_IdleTime+CPU_WorkingTime) * 100);
+      UART_TransmitChar(UART_CH_0, CPU_Load);
+      while(1)
+      {
+         
+      }
+   }
+#endif
+
+/**
+* @brief: This function is ISR of the Systick.
+*/
+void Systick_ISR(void)
+{
+   #if CALCULATE_CPU_LOAD
+      /* Increment Idle and Working Overflows */
+      if(CPU_IDLE == CPU_State)
+      {
+         CPU_IdleTime++;
+      }
+      else
+      {
+         CPU_WorkingTime++;
+      }
+   #endif
+   /* Increment overflows */
+   gu64_SYSTICK_Overflows++;
+   /* If overflows is equal to the Tick */
+   if(gu64_SYSTICK_Overflows == gu64_SYSTICK_OverflowsNeeded)
+   {
+      /* Set Overflows to zero. */
+      gu64_SYSTICK_Overflows = 0;
+      /* Increment system ticks. */
+      gu64_Ticks++;
+      #if CALCULATE_CPU_LOAD
+         /* Transmit CPU load when Ticks is equal to the Calculation Window. */
+         if(CPU_CALCULATION_WINDOW <= gu64_Ticks)
+         {
+            TransmitCPULoad();
+         }
+      #endif
+      /* Update The Tasks States after each system tick. */
+      SOS_UpdateStates();
+   }
+}
+
+/**
+* @brief: This function is the idle routine when no tasks are running.
+*/
+static void Idle_Task(void)
```

Comment:
in the idle task I guess it's better to make a flag for new tick, to make the system stuck in the idle task, not to go out and repeat the if conditions which will be the same output, stay in idle till a new tick is here then we see what to do.